### ISSUE TITLE
JackAudio Backend does not compile

### ISSUE THREAD

**Issue by StopMotionCuber** [librespot-org/librespot#216]
Hey everyone, I cannot compile librespot with JackAudio backend.
I'm compiling on a Raspberry Pi 1st Generation (arm).
I also tried to cross compile for it today on a x86 machine with the docker file provided.

Steps to reproduce:
execute `cargo build --no-default-features --release --features --jackaudio-backend` on a ARMv6 Machine
Expected Result:
librespot does compile without error
Actual Result:
Compilation of `librespot-playback v0.1.0` fails.
When running verbose, I'm getting this log:

```
    Compiling librespot-playback v0.1.0 (file:///home/pi/repositories/librespot/playback)
     Running `rustc --crate-name librespot_playback playback/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg 'feature="jack"' --cfg 'feature="jackaudio-backend"' -C metadata=d20cf74f3ad1be65 -C extra-filename=-d20cf74f3ad1be65 --out-dir /home/pi/repositories/librespot/target/release/deps -L dependency=/home/pi/repositories/librespot/target/release/deps --extern futures=/home/pi/repositories/librespot/target/release/deps/libfutures-39faefcd2f0d0247.rlib --extern jack=/home/pi/repositories/librespot/target/release/deps/libjack-f51c115881e0c066.rlib --extern byteorder=/home/pi/repositories/librespot/target/release/deps/libbyteorder-7fb62df717007fb7.rlib --extern librespot_audio=/home/pi/repositories/librespot/target/release/deps/liblibrespot_audio-82c88cd93111b51d.rlib --extern librespot_core=/home/pi/repositories/librespot/target/release/deps/liblibrespot_core-ac5aabdcc20cdd1f.rlib --extern log=/home/pi/repositories/librespot/target/release/deps/liblog-c53313b650fc9bb4.rlib --extern librespot_metadata=/home/pi/repositories/librespot/target/release/deps/liblibrespot_metadata-beafaa5e4aab6b03.rlib -L native=/home/pi/repositories/librespot/target/release/build/rust-crypto-b8055c8836ec1fd6/out`
error[E0308]: mismatched types
  --> playback/src/audio_backend/jackaudio.rs:18:35
   |
18 |   fn pcm_to_f32(sample: i16) -> f32 {
   |  ___________________________________^
19 | |     sample as f32 / 32768.0;
   | |                            - help: consider removing this semicolon
20 | | }
   | |_^ expected f32, found ()
   |
   = note: expected type `f32`
              found type `()`

error: aborting due to previous error

error: Could not compile `librespot-playback`.

Caused by:
  process didn't exit successfully: `rustc --crate-name librespot_playback playback/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg feature="jack" --cfg feature="jackaudio-backend" -C metadata=d20cf74f3ad1be65 -C extra-filename=-d20cf74f3ad1be65 --out-dir /home/pi/repositories/librespot/target/release/deps -L dependency=/home/pi/repositories/librespot/target/release/deps --extern futures=/home/pi/repositories/librespot/target/release/deps/libfutures-39faefcd2f0d0247.rlib --extern jack=/home/pi/repositories/librespot/target/release/deps/libjack-f51c115881e0c066.rlib --extern byteorder=/home/pi/repositories/librespot/target/release/deps/libbyteorder-7fb62df717007fb7.rlib --extern librespot_audio=/home/pi/repositories/librespot/target/release/deps/liblibrespot_audio-82c88cd93111b51d.rlib --extern librespot_core=/home/pi/repositories/librespot/target/release/deps/liblibrespot_core-ac5aabdcc20cdd1f.rlib --extern log=/home/pi/repositories/librespot/target/release/deps/liblog-c53313b650fc9bb4.rlib --extern librespot_metadata=/home/pi/repositories/librespot/target/release/deps/liblibrespot_metadata-beafaa5e4aab6b03.rlib -L native=/home/pi/repositories/librespot/target/release/build/rust-crypto-b8055c8836ec1fd6/out` (exit code: 101)
```

**Comment by StopMotionCuber** [comment]
Seems like this issue got introduced with [this](https://github.com/librespot-org/librespot/commit/f35f52cbf94548e11b0ad059d3cb754af18c5ec8) commit.
I have just manually undone this commit and it worked for me (I don't have experience with rust though so I do not know the clean way to resolve this issue)

**Comment by SirJson** [comment]
Have you tried removing that semicolon see if the JACK backend still works? Looks like a typo to me because that semicolon prevents returning the value pcm_to_f32 calculates.

... unfortunately I don't have JACK on my system installed to verify this myself but the code would at least compile.

**Comment by StopMotionCuber** [comment]
@SirJson Unfortunately it does not work for me. 
As soon as I'm trying to connect to the Raspberry Pi the jackaudio backend tries to connect to the jack server which is clearly running on my system.

Here the full log from journalctl:
```
May 14 20:07:22 StopMotionPi librespot[530]: INFO:librespot_core::session: Connecting to AP "gew1-accesspoint-b-n1bs.ap.spotify.com:4070"
May 14 20:07:23 StopMotionPi librespot[530]: INFO:librespot_core::session: Authenticated as "1131204546" !
May 14 20:07:23 StopMotionPi librespot[530]: INFO:librespot_core::session: Country: "DE"
May 14 20:07:23 StopMotionPi librespot[530]: INFO:librespot_playback::audio_backend::jackaudio: Using jack sink!
May 14 20:07:23 StopMotionPi librespot[530]: connect(2) call to /tmp/jack-109/default/jack_0 failed (err=No such file or directory)
May 14 20:07:23 StopMotionPi librespot[530]: attempt to connect to server failed
May 14 20:07:23 StopMotionPi librespot[530]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ClientError(FAILURE | SERVER_FAILED)', libcore/result.rs:945:5
May 14 20:07:23 StopMotionPi librespot[530]: stack backtrace:
May 14 20:07:24 StopMotionPi librespot[530]:    0:   0x9f53b7 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h4ef6490c84cef3d0
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
May 14 20:07:24 StopMotionPi librespot[530]:    1:   0x9ef113 - std::sys_common::backtrace::_print::hea150ee4f33bb5de
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/sys_common/backtrace.rs:71
May 14 20:07:24 StopMotionPi librespot[530]:    2:   0x9f3e27 - std::panicking::default_hook::{{closure}}::h6c4396ff5bd61d89
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/sys_common/backtrace.rs:59
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:380
May 14 20:07:24 StopMotionPi librespot[530]:    3:   0x9f3a0b - std::panicking::default_hook::h0f6cb16cf06a0198
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:396
May 14 20:07:24 StopMotionPi librespot[530]:    4:   0x9f42cb - std::panicking::rust_panic_with_hook::h2207195920ffa55f
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:576
May 14 20:07:24 StopMotionPi librespot[530]:    5:   0x9f412f - std::panicking::begin_panic::h8d43a09cf6d6e5f1
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:537
May 14 20:07:24 StopMotionPi librespot[530]:    6:   0x9f40b3 - std::panicking::begin_panic_fmt::hd9afc94159ee9735
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:521
May 14 20:07:24 StopMotionPi librespot[530]:    7:   0x9f404b - rust_begin_unwind
May 14 20:07:24 StopMotionPi librespot[530]:                        at libstd/panicking.rs:497
May 14 20:07:24 StopMotionPi librespot[530]:    8:   0xa4242b - core::panicking::panic_fmt::h0bd854df201d1baf
May 14 20:07:25 StopMotionPi librespot[530]:                        at libcore/panicking.rs:71
May 14 20:07:25 StopMotionPi librespot[530]:    9:   0x594817 - core::result::unwrap_failed::h4e0ffb7374851f60
May 14 20:07:25 StopMotionPi librespot[530]:   10:   0x5a272b - <librespot_playback::audio_backend::jackaudio::JackSink as librespot_playback::audio_backend::Open>::open::h2a9413d3b0a99960
May 14 20:07:25 StopMotionPi librespot[530]:   11:   0x5b8db3 - librespot_playback::audio_backend::mk_sink::hb1e79cd339d83e00
May 14 20:07:25 StopMotionPi librespot[530]:   12:   0x52363f - std::sys_common::backtrace::__rust_begin_short_backtrace::hbaacb6fe88345e64
May 14 20:07:25 StopMotionPi librespot[530]:   13:   0x506dc7 - std::panicking::try::do_call::h9144a2334eae091d
May 14 20:07:25 StopMotionPi librespot[530]:   14:   0xa036cb - __rust_maybe_catch_panic
May 14 20:07:25 StopMotionPi librespot[530]:                        at libpanic_unwind/lib.rs:102
May 14 20:07:25 StopMotionPi librespot[530]:   15:   0x507ff3 - <F as alloc::boxed::FnBox<A>>::call_box::hc3cc0f681c1033fe
May 14 20:07:25 StopMotionPi librespot[530]:   16:   0x9fa363 - std::sys::unix::thread::Thread::new::thread_start::hb1d1727b173caaf8
May 14 20:07:25 StopMotionPi librespot[530]:                        at /checkout/src/liballoc/boxed.rs:798
May 14 20:07:25 StopMotionPi librespot[530]:                        at libstd/sys_common/thread.rs:24
May 14 20:07:25 StopMotionPi librespot[530]:                        at libstd/sys/unix/thread.rs:90
```

Indeed `/tmp/jack-109/default/jack_0` does not exist on my machine.
The jack server is running at socket `/tmp/jack-1000/default/jack_0`.

According to the log messages the error has to be somewhere below [here](https://github.com/librespot-org/librespot/blob/8995f7609f5203172d21c790c02a1cd96a449148/playback/src/audio_backend/jackaudio.rs#L43).

**Comment by StopMotionCuber** [comment]
I replaced jackd1 with jackd2 and journalctl has different error messages for the jackd calls.
I'm not sure if this is relevant but I thought that additional information would be nice for someone debugging this.
```
May 14 22:43:32 StopMotionPi librespot[5626]: INFO:librespot: librespot deb240c (2018-05-01). Built on 2018-05-08. Build ID: 1o4tgwTs
May 14 22:43:38 StopMotionPi librespot[5626]: INFO:librespot_core::session: Connecting to AP "gew1-accesspoint-b-qkg1.ap.spotify.com:4070"
May 14 22:43:38 StopMotionPi librespot[5626]: INFO:librespot_core::session: Authenticated as "1131204546" !
May 14 22:43:38 StopMotionPi librespot[5626]: INFO:librespot_core::session: Country: "DE"
May 14 22:43:38 StopMotionPi librespot[5626]: INFO:librespot_playback::audio_backend::jackaudio: Using jack sink!
May 14 22:43:38 StopMotionPi librespot[5626]: Cannot connect to server socket err = No such file or directory
May 14 22:43:38 StopMotionPi librespot[5626]: Cannot connect to server request channel
May 14 22:43:38 StopMotionPi librespot[5626]: jack server is not running or cannot be started
May 14 22:43:38 StopMotionPi librespot[5626]: JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
May 14 22:43:38 StopMotionPi librespot[5626]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ClientError(FAILURE | SERVER_FAILED)', libcore/result.rs:945:5
May 14 22:43:38 StopMotionPi librespot[5626]: stack backtrace:
May 14 22:43:39 StopMotionPi librespot[5626]:    0:   0x9523b7 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h4ef6490c84cef3d0
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
May 14 22:43:39 StopMotionPi librespot[5626]:    1:   0x94c113 - std::sys_common::backtrace::_print::hea150ee4f33bb5de
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/sys_common/backtrace.rs:71
May 14 22:43:39 StopMotionPi librespot[5626]:    2:   0x950e27 - std::panicking::default_hook::{{closure}}::h6c4396ff5bd61d89
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/sys_common/backtrace.rs:59
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:380
May 14 22:43:39 StopMotionPi librespot[5626]:    3:   0x950a0b - std::panicking::default_hook::h0f6cb16cf06a0198
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:396
May 14 22:43:39 StopMotionPi librespot[5626]:    4:   0x9512cb - std::panicking::rust_panic_with_hook::h2207195920ffa55f
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:576
May 14 22:43:39 StopMotionPi librespot[5626]:    5:   0x95112f - std::panicking::begin_panic::h8d43a09cf6d6e5f1
May 14 22:43:39 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:537
May 14 22:43:40 StopMotionPi librespot[5626]:    6:   0x9510b3 - std::panicking::begin_panic_fmt::hd9afc94159ee9735
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:521
May 14 22:43:40 StopMotionPi librespot[5626]:    7:   0x95104b - rust_begin_unwind
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libstd/panicking.rs:497
May 14 22:43:40 StopMotionPi librespot[5626]:    8:   0x99f42b - core::panicking::panic_fmt::h0bd854df201d1baf
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libcore/panicking.rs:71
May 14 22:43:40 StopMotionPi librespot[5626]:    9:   0x4f1817 - core::result::unwrap_failed::h4e0ffb7374851f60
May 14 22:43:40 StopMotionPi librespot[5626]:   10:   0x4ff72b - <librespot_playback::audio_backend::jackaudio::JackSink as librespot_playback::audio_backend::Open>::open::h2a9413d3b0a99960
May 14 22:43:40 StopMotionPi librespot[5626]:   11:   0x515db3 - librespot_playback::audio_backend::mk_sink::hb1e79cd339d83e00
May 14 22:43:40 StopMotionPi librespot[5626]:   12:   0x48063f - std::sys_common::backtrace::__rust_begin_short_backtrace::hbaacb6fe88345e64
May 14 22:43:40 StopMotionPi librespot[5626]:   13:   0x463dc7 - std::panicking::try::do_call::h9144a2334eae091d
May 14 22:43:40 StopMotionPi librespot[5626]:   14:   0x9606cb - __rust_maybe_catch_panic
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libpanic_unwind/lib.rs:102
May 14 22:43:40 StopMotionPi librespot[5626]:   15:   0x464ff3 - <F as alloc::boxed::FnBox<A>>::call_box::hc3cc0f681c1033fe
May 14 22:43:40 StopMotionPi librespot[5626]:   16:   0x957363 - std::sys::unix::thread::Thread::new::thread_start::hb1d1727b173caaf8
May 14 22:43:40 StopMotionPi librespot[5626]:                        at /checkout/src/liballoc/boxed.rs:798
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libstd/sys_common/thread.rs:24
May 14 22:43:40 StopMotionPi librespot[5626]:                        at libstd/sys/unix/thread.rs:90
```

**Comment by StopMotionCuber** [comment]
Ignore my last 2 log messages, librespot was started as the wrong user.

**Comment by kingosticks** [comment]
That makes sense. Those paths did look like they contained userids (1000=pi, 109=system something).

**Comment by plietar** [comment]
Fixed by #222.
